packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +531,820 @@ class ShellRoute extends RouteBase {
+  StatefulShellRoute get route;
+
+  /// The GoRouterState associate with [route].
+  GoRouterState get routerState;
+
+  /// The index of the currently active [StatefulShellBranch].
+  ///
+  /// Corresponds to the index of the branch in the List returned from
+  /// branchBuilder of [StatefulShellRoute].
+  int get currentIndex;
+
+  /// The Navigator key of the current navigator.
+  GlobalKey<NavigatorState> get currentNavigatorKey;
+
+  /// Navigate to the current location of the shell navigator with the provided
+  /// index.
+  ///
+  /// This method will switch the currently active [Navigator] for the
+  /// [StatefulShellRoute] by replacing the current navigation stack with the
+  /// one of the route branch identified by the provided index. If resetLocation
+  /// is true, the branch will be reset to its initial location
+  /// (see [StatefulShellBranch.initialLocation]).
+  void goBranch({
+    required int index,
+  });
+}
+
+/// Builder for a custom container for shell route Navigators.
+typedef ShellNavigatorContainerBuilder = Widget Function(BuildContext context,
+    StatefulShellRouteState shellRouteState, List<Widget> children);
+
+/// Widget for managing the state of a [StatefulShellRoute].
+///
+/// Normally, this widget is not used directly, but is instead created
+/// internally by StatefulShellRoute. However, if a custom container for the
+/// branch Navigators is required, StatefulShellRouteController can be used in
+/// the builder or pageBuilder methods of StatefulShellRoute to facilitate this.
+/// The container is created using the provided [ShellNavigatorContainerBuilder],
+/// where the List of Widgets represent the Navigators for each branch.
+///
+/// Example:
+/// ```
+/// builder: (BuildContext context, StatefulShellRouteState state, Widget child) {
+///   return StatefulShellRouteController(
+///     shellRouteState: state,
+///     containerBuilder: (_, __, List<Widget> children) => MyCustomShell(shellState: state, children: children),
+///   );
+/// }
+/// ```
+abstract class StatefulShellRouteController extends StatefulWidget {
+  /// Constructs a [StatefulShellRouteController].
+  factory StatefulShellRouteController(
+          {required StatefulShellRouteState shellRouteState,
+          required ShellNavigatorContainerBuilder containerBuilder}) =>
+      _StatefulNavigationShell._(shellRouteState as _StatefulShellRouteState,
+          containerBuilder: containerBuilder);
+
+  const StatefulShellRouteController._();
+}
+
+/// Widget that manages and maintains the state of a [StatefulShellRoute],
+/// including the [Navigator]s of the configured route branches.
+///
+/// This widget acts as a wrapper around the builder function specified for the
+/// associated StatefulShellRoute, and exposes the state (represented by
+/// [StatefulShellRouteState]) to its child widgets with the help of the
+/// InheritedWidget [_InheritedStatefulNavigationShell]. The state for each route
```

Comment:
_InheritedStatefulNavigationShell is removed? Also I wonder why this is removed, it seems useful

Reply:
The _InheritedStatefulNavigationShell was removed due to the previous discussion and feedback around simplifying the builders of StatefulShellRoute, i.e. trying to avoid having to "wrap" the Widget built by the builders within _InheritedStatefulNavigationShell (and StatefulNavigationShell). Also, _InheritedStatefulNavigationShell was primarily used to provide access the StatefulShellRouteState, which is now instead passed directly to the builder functions.

Reply:
but the widget in the subtree may not be able to access the state unless you drill down the StatefulShellRouteState.

Reply:
Yes, that is true, but I don't believe that is the main use case. The primary use of StatefulShellRouteState would most likely be in the Widget implementing the shell. But we could certainly re-introduce the InheritedWidget and `StatefulShellRouteState.of`. The only problem is that you wouldn't be able to call it in the builder functions, or even in the shell Widget - only from Widgets of sub-routes of the shell route. And you have to make this clear in the documentation of `StatefulShellRouteState.of`. But it's certainly doable if you think we should do this.

Reply:
I was thinking about goBranch may be useful for widget in subtree. Or maybe that can be something handled in other class. This will further clean up the StatefulShellRouteState to be less mutable.

Reply:
Yes, I guess it might of course. I'll go ahead and re-introduce _InheritedStatefulNavigationShell and StatefulShellRouteState.of. But I'm not sure how this will further clean up StatefulShellRouteState though? We still need to hook up goBrach the same way as before.

Reply:
I think my suggestion was to add goBranch into StatefulNavigationShell, and expose the `StatefulNavigationShell.of` which returns StatefulShellRouteState. In the method it should use `findAncestorStateOfType`.

You can clean up  StatefulShellRouteState a bit.

Reply:
Problem is that (as of now), the state of StatefulNavigationShell would only be available to sub-routes of the shell route - and you really need it on a shell-level. So I don't see how that would work. 

On a different note, I've kinda of steered clear of findAncestorStateOfType, due to its "expensiveness" (i.e.O(N)), but I guess if you're certain you're relatively close to the "source" it could be ok.

Reply:
Yeah you are right, I was thinking if you may have a way to provide the referenece of the StatefulNavigationShellState to the shell builder, maybe through a global key or StatefulNavigationShellState getter? but that seems to be drilling down to a rabbit hole.... the reason I dislike put it on StatefulShellRouteState is that we have to store state of other branches and need to update it as it change.

> findAncestorStateOfType, due to its "expensiveness" (i.e.O(N))

I wouldn't be too worry about this if we don't expose property that requires for building widget on it (those properties should be on StatefuleShellRouteState). If we do that, the `.of` should only be called when handling onpress or other event call back, which wouldn't be too bad.